### DISCUSSION TITLE
Is there a way to access the field if it was hidden by a lexical variable with the same name?

### DISCUSSION THREAD

**Discussion by yuki-kimoto** [Ovid/Cor#75] [Q&A] [answered]
Is there a way to access the field if it was hidden by a lexical variable with the same name?

```
class MyClass {
  field $x;

  method foo () {
    my $x = 1;
    
    # How to access field $x?
    # In Java, this.x
  } 
}
```

**Comment by leonerd** [comment, marked as answer]
No. Don't do that.

**Reply by yuki-kimoto** [reply]
> No. Don't do that.

Is it intentional? Isn't it inconvenient not to have one?

Does it have something to do with Twigil(<code>$:x</code>) getting rejected?

**Reply by leonerd** [reply]
It's no more inconvenient than the inability to access the outer `$x` from inside `my $x; { my $x ... }` or `my $x; { our $x; ... }`. Variables are lexically scoped - declaring another variable of the same name will hide the first one. Fields are just variables. This applies.

**Reply by leonerd** [reply]
Twigils would indeed permit that. But it seems the intended direction of the project is not to use them.

**Reply by yuki-kimoto** [reply]
How would the following code be written in Collina?

```
# Current Perl Simple OO
sub foo ($self, $x, $y) {
  $self->{x} = $x;
  $self->{y} = $y:
}

# Corinna

```

**Reply by leonerd** [reply]
I'd probably write

```perl
method foo ($new_x, $new_y) {
  $x = $new_x;
  $y = $new_y;
}
```

**Reply by leonerd** [reply]
If instead you don't like that you might prefer to adopt the "prefix field names with an underscore" naming convention and end up with

```perl
class Example {
  field $_x;
  field $_y;
  method foo ($x, $y) {
    $_x = $x;
    $_y = $y;
  }
}
```

**Reply by yuki-kimoto** [reply]
Thanks. 

```
method foo ($new_x, $new_y) {
  $x = $new_x;
  $y = $new_y;
}
```

It means adjusting the name of the argument. OK. I will use this way.

```
class Example {
  field $_x;
  field $_y;
  method foo ($x, $y) {
    $_x = $x;
    $_y = $y;
  }
}
```

I wanted to do this, however, this seems that other complexity is added related to <code>:reader</code>, <code>:writer</code>, <code>:name</code>.

**Reply by yuki-kimoto** [reply]
I had an idea yesterday. so I write this.

```
class Example {
  field $x :reader :writer;
  field $y :reader :writer;
  method foo ($x, $y) {
    $self->set_x($x);
    $self->set_y($y);
  }
}
```

**Reply by Ovid** [reply]
That would work, so long as you agree to expose that data to the outside work via the readers and writers, but really it's a bad idea. Interfaces should be as small as possible. If you make the readers and writers available to consumers of that class but later realize you want to change the internals, you might be forever stuck with a bad interface. Don't design that interface as a small convenience for you. Design the interface so that _consumers_ have what they need and nothing more. Classes with too many readers and writers are a code smell.

So [as Paul said above](https://github.com/Ovid/Cor/discussions/75#discussioncomment-3478735), "don't do that."

**Reply by yuki-kimoto** [reply]
Note that I am not willing to try this way. 

Since there is no way in Corrina, I am writing it as a possible alternative.

If <code>$:x</code> is provided as <code>this.x</code> as Java, I will be satisfied.

**Reply by leonerd** [reply]
We haven't considered using twigils so far, as various people didn't like them in early design discussions, and the advocates couldn't find many reasons good enough to override that. If you want them to be seriously considered you'll have to make quite a strong case in their favour.

**Reply by yuki-kimoto** [reply]
Umm, I don't want Twigil. I want the same feature as <code>this.field</code> as Java.

```
field @field;
```

<code>$self~>@field</code>

<code>$self:@field</code>

<code>$self::::@field</code>

<code>@:field</code>

<code>this @field</code>

**Reply by leonerd** [reply]
The `$:field` syntax is called a twigil.

We did ponder on the idea of `field $thing` being similar to `our $thing`, in that duplicate declarations would in fact access the same one as before. That would permit `(field $x) = $x` in that situation. Problem is it's very messy - it then leaks out into the subsequent scope. We don't have lvalue `do {}` blocks so it's hard to contain it.

**Reply by yuki-kimoto** [reply]
> is it's very messy.

I can also accept other expression.

>  it then leaks out into the subsequent scope

Is it because <code>field</code> is a declaration? 

Isn't <code>this.x = 3</code> in Java related to the subsequent scope?

**Reply by leonerd** [reply]
> > is it's very messy.
> 
> I can also accept other expression.

Suggest one that isn't terrible.

> > it then leaks out into the subsequent scope
> 
> Is it because `field` is a declaration?

Yes.

> Isn't `this.x = 3` in Java related to the subsequent scope?

Not in the slightest. It doesn't declare anything new.
